node/npm_dependencies.ts

```diff
@@ -72,26 +124,39 @@ const getNPMSpecifiers = async (basePath: string, functions: string[], importMap
       return nftResolve(specifier, ...args)
     },
   })
-  const npmSpecifiers = new Set<string>()
+  const npmSpecifiers: { specifier: string; types?: string }[] = []
   const npmSpecifiersWithExtraneousFiles = new Set<string>()
 
-  reasons.forEach((reason, filePath) => {
-    const packageName = getPackageName(filePath)
+  for (const [packageJsonPath, reason] of reasons.entries()) {
+    // every dependency will have its `package.json` in `reasons` exactly once.
+    // by only looking at this file, we save us from doing duplicate work.
+    const isPackageJson = packageJsonPath.endsWith('package.json')
+    if (!isPackageJson) continue
 
-    if (packageName === undefined) {
-      return
-    }
+    const packageName = getPackageName(packageJsonPath)
+    if (packageName === undefined) continue
 
     const parents = [...reason.parents]
-    const isDirectDependency = parents.some((parentPath) => !parentPath.startsWith(`node_modules${path.sep}`))
+
+    const isDirectDependency = parents.some((parentPath) => {
+      if (!parentPath.startsWith(`node_modules${path.sep}`)) return true
+      // typically, edge functions have no direct dependency on the `package.json` of a module.
+      // it's the impl files that depend on `package.json`, so we need to check the parents of
+      // the `package.json` file as well to see if the module is a direct dependency.
+      const ownImports = [...(reasons.get(parentPath)?.parents ?? [])]
+      return ownImports.some((parentPath) => !parentPath.startsWith(`node_modules${path.sep}`))
+    })
 
     // We're only interested in capturing the specifiers that are first-level
     // dependencies. Because we'll bundle all modules in a subsequent step,
     // any transitive dependencies will be handled then.
     if (isDirectDependency) {
-      const specifier = getPackageName(filePath)
+      const specifier = getPackageName(packageJsonPath)
```

Comment:
I know that this already existed in the code, but I don't think we need to call `getPackageName` again. We can just use `packageName` from above?

Reply:
done in https://github.com/netlify/edge-bundler/pull/505/commits/a0a27f0cc7a1beee318a17642450fefe09afcb66